addons/test_mail/tests/test_mail_mail.py

```diff
@@ -60,6 +61,61 @@ def _reset_data(self):
         self.test_mail.write({'failure_reason': False, 'failure_type': False, 'state': 'outgoing'})
         self.test_notification.write({'failure_type': False, 'notification_status': 'ready'})
 
+    @users('admin')
+    def test_mail_mail_attachment_access(self):
+        mail = self.env['mail.mail'].create({
+            'body_html': 'Test',
+            'email_to': 'test@example.com',
+            'partner_ids': [(4, self.user_employee.partner_id.id)],
+            'attachment_ids': [
+                (0, 0, {'name': 'file 1', 'datas': 'c2VjcmV0'}),
+                (0, 0, {'name': 'file 2', 'datas': 'c2VjcmV0'}),
+                (0, 0, {'name': 'file 3', 'datas': 'c2VjcmV0'}),
+                (0, 0, {'name': 'file 4', 'datas': 'c2VjcmV0'}),
+            ],
+        })
+
+        def _patched_check(self, *args, **kwargs):
+            if self.env.is_superuser():
+                return
+            if any(attachment.name in ('file 2', 'file 4') for attachment in self):
+                raise AccessError('No access')
+
+        mail.invalidate_cache(ids=mail.ids)
+
+        new_attachment = self.env['ir.attachment'].create({
+            'name': 'new file',
+            'datas': 'c2VjcmV0',
+        })
+
+        with patch.object(type(self.env['ir.attachment']), 'check', _patched_check):
+            # Sanity check
+            self.assertEqual(mail.restricted_attachment_count, 2)
+            self.assertEqual(len(mail.unrestricted_attachment_ids), 2)
+            self.assertEqual(mail.unrestricted_attachment_ids.mapped('name'), ['file 1', 'file 3'])
+
+            # Add a new attachment
+            mail.write({
+                'unrestricted_attachment_ids': [Command.link(new_attachment.id)],
+            })
+            self.assertEqual(mail.restricted_attachment_count, 2)
+            self.assertEqual(len(mail.unrestricted_attachment_ids), 3)
+            self.assertEqual(mail.unrestricted_attachment_ids.mapped('name'), ['file 1', 'file 3', 'new file'])
+            self.assertEqual(len(mail.attachment_ids), 5)
+
+            # Remove an attachment
+            mail.write({
```

Comment:
We could try to reset command ((5, 0) and check other attachments are still there :)

Reply:
good idea :)